### PR TITLE
Enable writing Threads::Thread<T> for types T that are only movable.

### PR DESCRIPTION
**Pull request by bangerth**
Previously, it was not possible to write Threads::new_thread(&foo) or Threads::new_task(&foo)
for functions that return an object whose type is not copyable, but is movable. That's
because in a number of places in the internal machinery of the Threads::Thread and
Threads::Task classes, we didn't account for such types.

This patch fixes this. It is, admittedly a bit technical, but it was also fun to
get this figured out.

This is necessary for the patch I promised to write in #5880, and is also
going to be necessary for what @masterleinad offered to do there, so I'd
appreciate if someone looked at this in a reasonable time.

### CODE REVIEW

File: include/deal.II/base/thread_management.h [hunk 1]
```diff
@@ -692,16 +692,18 @@ namespace Threads
...
+      inline reference_type get ()
       {
         return value;
       }
 
-      inline void set (RT v)
+      inline void set (RT &&v)
       {
-        value = v;
+        value = std::move(v);
```
Comment (bangerth) [inline include/deal.II/base/thread_management.h]: This is the key change of it all. It makes sure that whatever the function called on a separate thread/task returns, is moved, not copied, to the location where we store the result of the function so that the place that created the thread/task can access it.

File: include/deal.II/base/thread_management.h [hunk 2]
```diff
@@ -1025,7 +1032,8 @@ namespace Threads
      * Get the return value of the function of the thread. Since this is only
      * available once the thread finishes, this implicitly also calls join().
      */
-    RT return_value ()
+    typename internal::return_value<RT>::reference_type
+    return_value ()
```
Comment (bangerth) [inline include/deal.II/base/thread_management.h]: And this (and the other place below for class `Task`) is the other key place: instead of returning the value returned by the thread/task function by value, we need to return a reference to it so that accessing functions don't have to move it away and can ask for it more than once.

File: include/deal.II/base/thread_management.h [hunk 3]
```diff
@@ -692,16 +692,18 @@ namespace Threads
     private:
       RT value;
     public:
+      typedef RT &reference_type;
+
       inline return_value () : value() {}
 
-      inline RT get () const
+      inline reference_type get ()
```
Comment (masterleinad) [inline include/deal.II/base/thread_management.h]: Is there anything that prevents this function from being `const`?

File: tests/base/thread_validity_14.cc [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+  // wait for the thread to return and query its value
+  deallog << t.return_value().value
+          << std::endl;
+
+  // we can't copy the return_value() object directly, but we can move
+  // it. do so and check that we still get the correct value. then
+  // also check that the value of the original return object has been
+  // reset in the move constructor/operator
+  X x = std::move(t.return_value());
```
Comment (masterleinad) [inline tests/base/thread_validity_14.cc]: Hmm... for me using this idiom looks a bit strange. Intuitively, I would not expect that I can modify the internal state of an object via `get_value`. What about letting `return_value()` return a `const &` and implementing a 
```
RT move_value()
{
  join();
  return std::move(return task_descriptor->ret_val.get()); 
}
```
function that can be used here instead?

File: include/deal.II/base/thread_management.h (hunk 3 above)
Comment (bangerth) [reply include/deal.II/base/thread_management.h]: Yes. If you make it `const`, the return type also has to be a `const` reference and you can't `std::move` the object that is returned.

**Comment by bangerth** [comment]
@masterleinad -- I came up with this design because as explained in a comment, you can't make the `get()` function `const` or it becomes unusable if the object is only movable. You work around this by introducing a second function, `move_value()` because you say that it's surprising that you can modify the stored value, but I think that doesn't reduce any surprise: if you *thought* that the returned object is unmodifiable, then the existence of `move_value()` should come as as big a surprise as the fact that you can write `std::move(get())`.

The way I tend to think is that the object just stores whatever the function returned, somewhere. I don't see a particularly good reason why that object should a-priori be `const` -- it's just a variable, stored somewhere. If you don't want it modifiable, then mark the surrounding object `const`.

In other words, I still like my design :-)

File: include/deal.II/base/thread_management.h (hunk 1 above)
Comment (drwells) [reply include/deal.II/base/thread_management.h]: Is the call to `std::move` necessary? `v` is already an rvalue reference at that point.

File: include/deal.II/base/thread_management.h (hunk 3 above)
Comment (masterleinad) [reply include/deal.II/base/thread_management.h]: Good point.

**Comment by masterleinad** [comment]
OK, I don't want to insist. Still, I let someone else push the merge button.

**Comment by masterleinad** [comment]
@dealii/dealii Can someone please have a look?

**Comment by bangerth** [comment]
Ping?

**Comment by Rombur** [comment]
I think the design is fine. If there is no `const`, there is no reason to expect that internal is not modified.
